### PR TITLE
Handle grayscale images

### PR DESCRIPTION
**Pull request by aksdb**
Up until now, it was up to libvips how the colors were handled. This is relevant and a problem for cases where we apply the existing image to a new background: embedding (putting the image on a larger surface with a specific background color) and flattening (removing the alpha channel and replace it with a specific color).

If there were only 1 or 2 channels (grayscale), libvips simply used the first channel of the specified RGB values (i.e. red), the rest was discarded. In one case, however, it even failed: when embedding with transparency on.

To fix both, I introduced some kind of "smart" conversion: when a grayscale image is embedded or flattened to a grayscale background (R=G=B), everything's fine. We can safely just use the red channel. If, however, we embed on a color (!(R=G=B)), we convert the image to RGB first (adding the missing channels). That way we silently honor the wish for the given color.

I think this is the best handling, since if someone requests the image to be on a specifically colored background, there is likely a reason for this and it does not seem nice to silently ignore this and keep the image grayscale. If someone wants color, it's apparently not intended that the image was grayscale in the first place.

### CODE REVIEW

File: transformation_test.go [hunk 1]
```diff
@@ -104,3 +104,127 @@ func TestImageTransformation_Resize(t *testing.T) {
...
+		if err := imageTrans.Embed(EmbedOptions{
+			Width:      200,
+			Height:     200,
+			Extend:     ExtendBackground,
+			Background: Color{R: 255, G: 0, B: 0},
+		}); err != nil {
+			t.Fatalf("embed returned unexpected error: %v", err)
+		}
+		if imageTrans.Metadata().Channels != 4 {
+			t.Fatalf("image should still have four channels now")
```
Comment (stefan-scheidewig) [inline transformation_test.go]: I guess "still" is wrong here as the number of channels went from 2 to 4.

Comment (aksdb) [reply transformation_test.go]: Copy & Paste .... dammit. Thanks :-)

File: transformation_test.go [hunk 2]
```diff
@@ -104,3 +104,127 @@ func TestImageTransformation_Resize(t *testing.T) {
...
+			t.Fatalf("cannot load image: %v", err)
+		}
+		if imageTrans.Metadata().Channels != 2 {
+			t.Fatalf("source image has unexpected number of channels")
+		}
+		if err := imageTrans.Flatten(Color{R: 255, G: 255, B: 255}); err != nil {
+			t.Fatalf("flatten returned unexpected error: %v", err)
+		}
+		if imageTrans.Metadata().Channels != 1 {
+			t.Errorf("image should still have just one channel")
```
Comment (stefan-scheidewig) [inline transformation_test.go]: The alpha channel is gone so there is just one grayscale channel left.

```suggestion
			t.Errorf("image should now have just one channel")
```

File: transformation_test.go [hunk 3]
```diff
@@ -104,3 +104,127 @@ func TestImageTransformation_Resize(t *testing.T) {
...
+		if err := imageTrans.Flatten(Color{R: 255, G: 0, B: 0}); err != nil {
+			t.Fatalf("flatten returned unexpected error: %v", err)
+		}
+		if imageTrans.Metadata().Channels != 3 {
+			t.Errorf("image should have three channels now")
+		}
+		if out, err := imageTrans.Save(SaveOptions{}); err != nil {
+			t.Errorf("cannot save image: %v", err)
+		} else {
+			Write("testdata/transformation_flatten_bwa_out.png", out)
```
Comment (stefan-scheidewig) [inline transformation_test.go]: Should be different than the output of the other test else you cannot compare it manually.

```suggestion
			Write("testdata/transformation_flatten_bwa_on_red_out.png", out)
```

**Comment by aksdb** [comment]
@stefan-scheidewig Your findings should be fixed in 730d0b4b.

**Review by stefan-scheidewig — APPROVED** [review]

**Pull request merged**